src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+                    var tbv = new TableValueBuilder
+                    {
+                        {tvr.Read((int) DocumentsTable.LoweredKey, out size), size},
+                        newEtagBigEndian,
+                        {tvr.Read((int) DocumentsTable.Key, out size), size},
+                        {data.BasePointer, data.Size},
+                        {(byte*) pChangeVector, sizeof(ChangeVectorEntry) * changeVector.Length},
+                        modifiedTicks,
+                        (int) (*(DocumentFlags*) tvr.Read((int) DocumentsTable.Flags, out size) | DocumentFlags.HasAttachments),
+                        transactionMarker
+                    };
+
+                    table.Update(tvr.Id, tbv);
+                }
+
+                _documentDatabase.Metrics.DocPutsPerSecond.MarkSingleThreaded(1);
+                _documentDatabase.Metrics.BytesPutsPerSecond.MarkSingleThreaded(data.Size);
+
+                context.Transaction.AddAfterCommitNotification(new AttachmentChange
+                {
+                    Etag = newEtag,
+                    CollectionName = collectionName.Name,
+                    Key = documentId,
+                    Name = name,
+                    Type = DocumentChangeTypes.PutAttachment,
+                    IsSystemDocument = collectionName.IsSystem,
+                });
+            }
+        }
+
+        private void ThrowDocumentNotFound(DocumentsOperationContext context, Slice loweredKey, string documentId, string name)
+        {
+            if (_conflictCount > 0)
+            {
+                try
+                {
+                    ThrowOnDocumentConflict(context, loweredKey);
+                }
+                catch (DocumentConflictException e)
+                {
+                    throw new InvalidOperationException($"Cannot put attachment {name} on a document '{documentId}' with a conflict.", e);
+                }
+            }
+            throw new InvalidOperationException($"Cannot put attachment {name} on a not exist document '{documentId}'.");
+        }
+
+        private IEnumerable<LazyStringValue> GetAttachmentsInternal(DocumentsOperationContext context, Slice startSlice)
+        {
+            var table = context.Transaction.InnerTransaction.OpenTable(AttachmentsSchema, AttachmentsMetadataSlice);
+            foreach (var sr in table.SeekForwardFrom(AttachmentsSchema.Indexes[AttachmentsKeyIndexSlice], startSlice, true))
+            {
+                foreach (var tvr in sr.Results)
+                {
+                    yield return TableValueToKey(context, (int)AttachmentsTable.Name, ref tvr.Reader);
+                }
+            }
+        }
+
+        public IEnumerable<LazyStringValue> GetAttachmentNamesForDocument(DocumentsOperationContext context, LazyStringValue lowerDocumentId)
+        {
+            Slice startSlice;
+            // TODO: Should we use here an enumerator or reutrn an array and dispose?
```

Comment:
Done, but it means that we may load the attachments in cases that we are not needing them (if we have a case that load without sending to the user).